### PR TITLE
[issue 161] Better message when missing contract on function call

### PR DESCRIPTION
**Pull request by djrtwo**
### What was wrong?
Issue [161](https://github.com/pipermerriam/web3.py/issues/161)

### How was it fixed?
Added checks for type of error and added specific message when missing contract.
Added tests for both types of errors:
- Wrong function call on contract
- Missing contract at address entirely

#### Cute Animal Picture

Cute animal picture

### CODE REVIEW

File: web3/contract.py [hunk 1]
```diff
@@ -778,14 +778,22 @@ def call_contract_function(contract,
...
         # eth-abi-utils
-        msg = (
-            "Could not decode contract function call {} return data {} for "
-            "output_types {}".format(
-                function_name,
-                return_data,
-                output_types
+        empty_strings = ["0x", b"0x", ""]
+        if (return_data in empty_strings
+                and contract.web3.eth.getCode(contract.address) in empty_strings):
```
Comment (pipermerriam) [inline web3/contract.py]: *nitpick*

I have a somewhat gigantic aversion to multi-line `if` statements.  Can you reformat this to something like the following or an alternative that doesn't split the `if` statement lines?

```python
is_missing_code_error = (
    return_data in empty_strings and
    contract.web3.eth.getCode(contract_address) in empty_strings
)
if is_missing_code_error:
    ...
```

File: web3/contract.py [hunk 2]
```diff
@@ -778,14 +778,22 @@ def call_contract_function(contract,
     except DecodingError as e:
         # Provide a more helpful error message than the one provided by
         # eth-abi-utils
-        msg = (
-            "Could not decode contract function call {} return data {} for "
-            "output_types {}".format(
-                function_name,
-                return_data,
-                output_types
+        empty_strings = ["0x", b"0x", ""]
```
Comment (pipermerriam) [inline web3/contract.py]: Can you move this to the top level of the module so it doesn't have to be re-allocated everytime.  Also, rename to cap-cased and something bytecode specific like `EMPTY_BYTECODE_VALUES`

File: web3/contract.py (hunk 1 above)
Comment (djrtwo) [inline web3/contract.py]: not a problem. I was struggling with figuring out how python standards wanted me to format it. Good approach with the boolean assignment :)

File: web3/contract.py (hunk 2 above)
Comment (djrtwo) [inline web3/contract.py]: 👍

**Comment by djrtwo** [comment]
@pipermerriam submit changes. ready for review

**Comment by kivo360** [comment]
I'm getting the same error. The main problems is that I see the abi of the contract, yet I can't access the method.

image

image


As you can see the function is inside of the contract, but it's not registering in web3. I see that it says the chain is not synced correctly, but I don't believe that is the main problem.

**Comment by carver** [comment]
@kivo360 you will need your chain to be synced, otherwise the contract won't have been deployed in your node yet. Also, please open a new issue, otherwise your question is likely to get lost. Finally, you may find that ethereum.stackexchange.com has the best response time for general questions, until they have confirmed that it's a web3.py bug.

**Comment by mamoona-aslam** [comment]
@kivo360  did u solve it??/

**Pull request merged**